[mono] Unify invoke code with coreclr

BrzVlad opened the pull request:
Unification of invoke paths enables use of IL Emit Invoke also on mono.

InternalInvoke now receives arguments byref, enabling support for future APIs making use of this.

Fixes https://github.com/dotnet/runtime/issues/68928
Fixes https://github.com/dotnet/runtime/issues/67457

BrzVlad committed:
[mono] Add icalls to be used by new invoke machinery

BrzVlad committed:
[mono] Fix IsInstanceOfType check

BrzVlad committed:
[mono] Reuse most of the managed invoke path used by coreclr

BrzVlad committed:
[mono] Implement new version of InternalInvoke icall

This version doesn't receive a MonoArray of params as objects, rather a simple array of byrefs. There is also no need for special handling for nullables. The runtime invoke wrapper receives a boxed nullable, this conversion happens in managed code using ReboxFromNullable and ReboxToNullable. This change might have some side effects on other API making use of the runtime invoke wrapper!

BrzVlad committed:
[tests] Enable reflection invoke tests on mono

BrzVlad committed:
[mono] Throw NRE for a null byref parameter

Behavior was changed recently.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -1795,6 +1799,41 @@ ves_icall_RuntimeTypeHandle_IsInstanceOfType (MonoQCallTypeHandle type_handle, M
 
+MonoObjectHandle
+ves_icall_RuntimeMethodHandle_ReboxToNullable (MonoObjectHandle obj, MonoQCallTypeHandle type_handle, MonoError *error)
+{
```
These two can return a value using a MonoObjectHandleOnStack.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -3360,13 +3399,63 @@ ves_icall_RuntimeMethodInfo_GetGenericArguments (MonoReflectionMethodHandle ref_
+{
+	gpointer result = NULL;
+	error_init (error);
+again:
```
No need to init the error.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -3436,19 +3524,87 @@ ves_icall_InternalInvoke (MonoReflectionMethodHandle method_handle, MonoObjectHa
+			obj = MONO_HANDLE_RAW (obj_h);
+			g_assert (obj); /*maybe we should raise a TLE instead?*/
+			if (m_class_is_valuetype (m->klass))
+				o = (MonoObject *)mono_object_unbox_internal ((MonoObject *)obj);
```
This cast is not needed.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -3436,19 +3524,87 @@ ves_icall_InternalInvoke (MonoReflectionMethodHandle method_handle, MonoObjectHa
+			MonoObjectHandle obj_h = mono_object_new_handle (m->klass, error);
+			goto_if_nok (error, return_null);
+			obj = MONO_HANDLE_RAW (obj_h);
+			g_assert (obj); /*maybe we should raise a TLE instead?*/
```
Its ok to assert, obj==null can't happen.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -3436,19 +3524,87 @@ ves_icall_InternalInvoke (MonoReflectionMethodHandle method_handle, MonoObjectHa
+			MonoObjectHandle obj_h = mono_value_box_handle (m->klass, obj, error);
+			goto_if_nok (error, return_null);
+			obj = MONO_HANDLE_RAW (obj_h);
+		}
```
Shouldn't this set 'o' ?

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -3436,19 +3524,87 @@ ves_icall_InternalInvoke (MonoReflectionMethodHandle method_handle, MonoObjectHa
+
+		if (!obj) {
+			MonoObjectHandle obj_h = mono_object_new_handle (m->klass, error);
+			goto_if_nok (error, return_null);
```
Might be better to use the non _handle functions and MONO_HANDLE_PIN () in this function.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -3436,19 +3524,87 @@ ves_icall_InternalInvoke (MonoReflectionMethodHandle method_handle, MonoObjectHa
+				goto_if_nok (error, return_null);
+				MonoObject* nullable = MONO_HANDLE_RAW (nullable_h);
+
+				MonoObjectHandle boxed_h = mono_value_box_handle (m_class_get_cast_class (m->klass), obj, error);
```
Might be easier to use mono_nullable_init_unboxed ().

BrzVlad committed:
[mono] Avoid loading reflection invoke machinery during startup

Aside from perf optimization, this allows the usage of LocalAppContextSwitches for invoke tests. Before this change, invoke machinery was initialized before the appcontext properties were set, failing to detect ForceInterpretedInvoke and ForceEmitInvoke properties. CoreCLR also hardcodes string type param.

srxqds commented:
hi, this will merge to rc1?

BrzVlad committed:
[mono] Remove nullable handling case

This API already receives a boxed nullable now.

BrzVlad committed:
[mono] Disable inlining into Invoke stubs

CoreCLR achieves this via NextCallReturnAddress. Add another intrinsic to be used for mono.

BrzVlad committed:
[mono] Fix handling of nullables when invoking from debugger

BrzVlad committed:
[mono] Update code for dyn runtime invokes and llvmonly invoke

Remove special handling for nullables and throwing of NullByRefReturnException.

marek-safar commented:
Could you look into fixing this for DynamicMethod as well?

BrzVlad commented:
> Could you look into fixing this for DynamicMethod as well?

Not sure what you mean by this. Most of the code for invoking a DynamicMethod is shared now.

BrzVlad commented:
> > Not sure what you mean by this. Most of the code for invoking a DynamicMethod is shared now.
> 
> Code like
> 
> https://github.com/dotnet/runtime/blob/561b0974793ba91936f151eab07875938e853694/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs#L160

@marek-safar It's not quite clear to me why that code is like that. Seems like duplicated code from `RuntimeMethodInfo` which mono calls directly instead. I'd rather leave it out of this PR anyway.

AaronRobinsonMSFT commented on src/mono/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.Mono.cs:
```diff
@@ -39,6 +39,11 @@ private static EqualityComparer<T> CreateComparer()
             {
                 return (EqualityComparer<T>)(object)(new ByteEqualityComparer());
             }
+            else if (t == typeof(string))
```
There is a comment on line 34 about "METHOD-TO-IR.C should be kept in sync. Is that still valid?

BrzVlad commented on src/mono/mono/mini/mini-amd64.c:
```diff
@@ -2544,7 +2544,7 @@ mono_arch_emit_setret (MonoCompile *cfg, MonoMethod *method, MonoInst *val)
 	MonoMethodSignature *sig;
 	CallInfo *cinfo;
-	int nstack_args, nullable_area;
+	int nstack_args;
```
@vargaz This PR removes special handling of nullables from the runtime invoke paths. The corresponding change in runtime invoke wrapper is made in `emit_invoke_call` in commit `[mono] Implement new version of InternalInvoke icall`. There is a risk for breaking changes with embedders but I don't think it is worthwhile to maintain separate versions of invoke wrappers and code.

lambdageek replied:
> There is a risk for breaking changes with embedders 

How does this affect embedders? Do they need to manually box nullable arguments before calling `mono_runtime_invoke`? That seems like a pretty big change

BrzVlad replied:
@lambdageek So consider the case where we do a runtime invoke into a method receiving a nullable argument. This method will expect to receive a nullable structure. The old Invoke API would receive an array of objects, so the caller of the Invoke API, when converting the argument into an object it will result a boxed value type. The invoke wrapper used to receive this boxed value type and it would unbox it into a nullable structure to pass it along. The new API would pass directly a ref to the nullable structure to the `InternalInvoke`, so the runtime invoke wrapper doesn't need to do anything else.

Now considering the case for embedders, this behavior of unboxing a valuetype into a nullable structure doesn't really seem to be documented and I figure it was just something for us to easily support the InternalInvoke behavior. The documentation of `mono_runtime_invoke` states that it should receive a `MonoObject*` for objects and a pointer to the value type for valuetypes, so this implementation still follows the documentation for the invokes. It just stops supporting, the apparently undocumented, behavior of allowing to pass a boxed valuetype as an argument for a method receiving a nullable.

BrzVlad replied:
I think we could also do the conversion from boxed vt to unboxed nullable only for `mono_runtime_invoke` and `mono_runtime_invoke_array`, just to preserve behavior

vargaz replied:
That should work.

vargaz replied:
So the nullable behavior is going to be handled in the invoke wrapper, so the dyn call code doesn't need these special cases anymore ?

BrzVlad committed:
[mono] Remove old invoke code

Which was still used only by mono_runtime_invoke_array

BrzVlad committed:
[mono] Move invoke code back in object.c

vargaz commented on src/mono/mono/metadata/class-setup-vtable.c:
```diff
@@ -1065,7 +1065,7 @@ handle_dim_conflicts (MonoMethod **vtable, MonoClass *klass, GHashTable *conflic
 		for (l = entries; l; l = l->next) {
-			if (l->data) {
+			if (l->data && l->data != impl) {
 				nentries ++;
```
Is this part of the PR ?

BrzVlad replied:
looks like rebase leftover

BrzVlad replied:
it was actually a bug that I forgot that I fixed

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -1795,6 +1799,48 @@ ves_icall_RuntimeTypeHandle_IsInstanceOfType (MonoQCallTypeHandle type_handle, M
+		return;
+	}
+
+	MonoObjectHandle res = mono_object_new_handle (klass, error);
```
This could use the newer way of handles, i.e. using raw object references and MONO_HANDLE_PIN ().

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -1795,6 +1799,48 @@ ves_icall_RuntimeTypeHandle_IsInstanceOfType (MonoQCallTypeHandle type_handle, M
+
+	gpointer vbuf = mono_object_unbox_internal (MONO_HANDLE_RAW (obj));
+	MonoObjectHandle res = mono_nullable_box_handle (vbuf, klass, error);
+
```
Ditto.

BrzVlad committed:
[mono] Add back support for the embedding api

With same behavior as before, except we have to handle conversions between nullable and boxedvt ourserlves, since it is no longer done by the runtime invoke wrapper.

BrzVlad committed:
[mono] Extract invoke code to be reused by mono_runtime_invoke_array

BrzVlad committed:
Disable some warnings

BrzVlad committed:
[mono] Simplify nullable ctor case

Nullable<T> only has 1 param constructor that creates a boxed vtype

BrzVlad committed:
[mono] Don't trigger ambiguous method exception when having duplicate method overrides

Seems to be possible via reflection, ex System.Reflection.Emit.Tests.MethodBuilderDefineMethodOverride.DefineMethodOverride_CalledTwiceWithSameBodies_Works

BrzVlad commented:
/azp run runtime-wasm
/azp run runtime-extra-platforms

BrzVlad commented:
/azp run runtime-extra-platforms

BrzVlad committed:
Disable tests on wasm aot since they rely on stacktraces

BrzVlad commented:
/azp run runtime-extra-platforms

BrzVlad commented:
/azp run runtime-wasm

BrzVlad commented:
@vargaz @lambdageek I'm not aware of any issues with this PR and I also updated the code to preserve behavior with embedders when passing nullables. Could I get a final review on this ?

vargaz commented on src/mono/System.Private.CoreLib/src/System/RuntimeMethodHandle.cs:
```diff
@@ -90,7 +90,25 @@ internal bool IsNullHandle()
-        internal static object? ReboxFromNullable(object? src) => src;
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal static extern void ReboxFromNullable (object? src, ObjectHandleOnStack res);
+
```
This could be private ?

vargaz commented on src/mono/System.Private.CoreLib/src/System/RuntimeMethodHandle.cs:
```diff
@@ -90,7 +90,25 @@ internal bool IsNullHandle()
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal static extern void ReboxToNullable (object? src, QCallTypeHandle destNullableType, ObjectHandleOnStack res);
+
```
Ditto.

vargaz commented on src/mono/mono/metadata/class-setup-vtable.c:
```diff
@@ -1065,7 +1065,7 @@ handle_dim_conflicts (MonoMethod **vtable, MonoClass *klass, GHashTable *conflic
 		for (l = entries; l; l = l->next) {
-			if (l->data) {
+			if (l->data && l->data != impl) {
 				nentries ++;
```
How is this related to the PR ?

BrzVlad replied:
I stumbled across this bug after changing the invoke machinery. I don't remember exactly why the test was passing before.

vargaz commented on src/mono/System.Private.CoreLib/src/System/RuntimeMethodHandle.cs:
```diff
@@ -90,7 +90,25 @@ internal bool IsNullHandle()
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal static extern void ReboxToNullable (object? src, QCallTypeHandle destNullableType, ObjectHandleOnStack res);
+
+        internal static object ReboxFromNullable (object? src)
```
Why are these methods in RuntimeMethodHandle ?

BrzVlad replied:
Just because they live in the same class on CoreCLR side

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -2089,31 +2136,9 @@ ves_icall_RuntimeFieldInfo_SetValueInternal (MonoReflectionFieldHandle field, Mo
 			g_assert (!gclass->context.class_inst->is_open);
 
-			if (mono_class_is_nullable (mono_class_from_mono_type_internal (type))) {
-				MonoClass *nklass = mono_class_from_mono_type_internal (type);
```
Why is this removed ?

BrzVlad replied:
The nullable boxing is no longer done here but from managed code via the Rebox{To/From}Nullable, from `RuntimeType.CheckValue`

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -6138,6 +6163,25 @@ ves_icall_System_RuntimeType_CreateInstanceInternal (MonoQCallTypeHandle type_ha
+/* Only used for value types */
+MonoObjectHandle
+ves_icall_System_RuntimeType_AllocateValueType (MonoQCallTypeHandle type_handle, MonoObjectHandle value_h, MonoError *error)
+{
```
This could return a result using a MonoObjectHandleOnStack.

marek-safar commented on src/mono/System.Private.CoreLib/src/System/RuntimeMethodHandle.cs:
```diff
@@ -90,7 +90,24 @@ internal bool IsNullHandle()
-        // Temporary placeholder until Mono adds support for supporting boxing true Nullables.
-        internal static object? ReboxFromNullable(object? src) => src;
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        private static extern void ReboxFromNullable (object? src, ObjectHandleOnStack res);
```
Are these same as https://github.com/dotnet/runtime/blob/main/src/mono/System.Private.CoreLib/src/System/Nullable.Mono.cs#L17-L39 ?

BrzVlad commented:
can we merge this ?
